source/common/websocket/codec.h

```diff
@@ -0,0 +1,179 @@
+                      uint32_t masking_key, std::vector<uint8_t>& output);
+
+  // Prepend the WebSocket frame header into the buffer.
+  // @param flags_and_opcode supplies the first fixed byte with flags and opcode.
+  // @param length supplies the WebSocket data frame length.
+  // @param is_masked supplies whether the frame is masked ot not.
+  // @param masking_key supplies the masking key if the is_masked is true.
+  // @param buffer the complete buffer with the payload.
+  void prependFrameHeader(uint8_t flags_and_opcode, uint64_t length, bool is_masked,
+                          uint32_t masking_key, Buffer::Instance& buffer);
+};
+
+enum class State {
+  // Decoding the first byte. Waiting for decoding the final frame flag (1 bit)
+  // and reserved flags (3 bits) and opcode (4 bits) of the WebSocket data frame.
+  FhFlagsAndOpcode,
+  // Decoding the second byte. Waiting for decoding the mask flag (1 bit) and
+  // length/length flag (7 bit) of the WebSocket data frame.
+  FhMaskFlagAndLength,
+  // Waiting for decoding the extended length of the frame if length read previously
+  // is either 126 or 127. Respectively 2 bytes or 8 bytes will be decoded from the
+  // WebSocket data frame.
+  FhExtendedLength,
+  // Waiting for decoding the masking key (4 bytes) only if the mask bit is set.
+  FhMaskingKey,
+  // Waiting for decoding the payload (both extension data and application data).
+  Payload,
+};
+
+class FrameInspector {
+public:
+  // Inspects the given buffer with WebSocket data frames and updates the frame count.
+  // Invokes visitor callbacks for each frame in the following sequence:
+  // ----------------------------------------------------------------------------------
+  //  frameStart frameMaskFlag frameMaskingKey? frameDataStart frameData* frameDataEnd
+  // ----------------------------------------------------------------------------------
+  // If frameStart returns false, then the inspector aborts.
+  // Returns the increase in the frame count.
+  uint64_t inspect(const Buffer::Instance& input);
+
+  // Returns the current frame count, corresponding to the request/response
+  // message count. Counter is incremented on a frame start.
+  uint64_t frameCount() const { return count_; }
+
+  // Returns the current state in the frame parsing.
+  State state() const { return state_; }
+
+  virtual ~FrameInspector() = default;
+
+protected:
+  virtual bool frameStart(uint8_t) { return true; }
+  virtual void frameMaskFlag(uint8_t mask_and_length) {
+    masking_key_length_ = (mask_and_length & 0x80) ? MASKING_KEY_LENGTH : 0;
+    length_ = mask_and_length & 0x7F;
+  }
+  virtual void frameMaskingKey() {}
+  virtual void frameDataStart() {}
+  virtual void frameData(uint8_t*, uint64_t) {}
+  virtual void frameDataEnd() {}
+
+  State state_{State::FhFlagsAndOpcode};
+  uint64_t length_{0};
+  uint8_t length_of_extended_length_{0};
+  uint32_t masking_key_{0};
+  uint8_t masking_key_length_{0};
+  uint64_t count_{0};
```

Comment:
what does `count_` represent?

Reply:
That is the total frames count when identified when inspect was called. I have renamed it to `total_frames_count_`.